### DISCUSSION TITLE
Binary releases?

### DISCUSSION THREAD

**Discussion by DannyBen** [open2b/scriggo#875] [Q&A] [answered]
Is there a plan to provide compiled binary releases on the GitHub releases page?

**Comment by gazerro** [comment, marked as answer]
We could also provide the binary releases of the scriggo command but note that if you have Go installed on your machine you can install it with

```shell
go install github.com/open2b/scriggo/cmd/scriggo@latest
```

If Go is not installed on the machine, the scriggo command would still be useful for running templates. 

@DannyBen is that what you ask or do you expect something different? Thanks.

**Reply by DannyBen** [reply]
Well - I am not a Go developer, so personally I have no use for scriggo in go code. As I understand, it is available as a command line utility, that renders the templates, exactly like Ruby's `erb`. If my understanding is correct, then the answer is yes: I would love to have the binary available (as I see many go/rust packages do) so that people who just want the CLI, can download it without hassle.

I believe the go developers are blessed with the grace of [goreleaser](https://github.com/goreleaser/goreleaser), which automates this process.

**Reply by gazerro** [reply]
To execute `erb` you should have Ruby installed or there is a binary? With a binary it is not possible to use other builtins other than those compiled into the binary.

Despite this limitation, I think it is a good idea to provide the binaries.

**Reply by DannyBen** [reply]
The `erb` executable comes with Ruby's standard installation.

```
$ erb --help
print this help
erb [switches] [var=value...] [inputfile]
  -x               print ruby script
  -n               print ruby script with line number
  -v               enable verbose mode
  -d               set $DEBUG to true
  -r library       load a library
  -E ex[:in]       set default external/internal encodings
  -U               set default encoding to UTF-8
  -T trim_mode     specify trim_mode (0..2, -)
  -P               disable ruby code evaluation for lines beginning with %
  var=value        set variable
```

I was hoping that the binary `scriggo` command, can be executed with a filename parameter (that will hold a template like [the one in the docs](https://scriggo.com/templates), and it will render it. No? 

If this is not a use case provided by scriggo, then I have wasted your time and I am sorry.
If it *is*, then it will be great.

**Reply by gazerro** [reply]
The scriggo serve command starts a web server and serves the template in the current directory.

```
$ scriggo help serve
usage: scriggo serve [-S n] [--metrics]

Serve runs a web server and serves the template rooted at the current
directory. It is useful to learn Scriggo templates.

It renders HTML and Markdown files based on file extension.

For example:

    http://localhost:8080/article

it renders the file 'article.html' as HTML if exists, otherwise renders the
file 'article.md' as Markdown.

Serving a URL terminating with a slash:

    http://localhost:8080/blog/

it renders 'blog/index.html' or 'blog/index.md'.

Markdown is converted to HTML with the Goldmark parser with the options
html.WithUnsafe and parser.WithAutoHeadingID.

Templates are automatically rebuilt when a file changes.

The -S flag prints the assembly code of the served file and n determines the
maximum length, in runes, of disassembled Text instructions

    n > 0: at most n runes; leading and trailing white space are removed
    n == 0: no text
    n < 0: all text

The --metrics flags prints metrics about execution time.
```

Would you like to be able to run the template and print it to the standard output?

> If this is not a use case provided by scriggo, then I have wasted your time and I am sorry.

No waste of time, we are very interested in different use cases and I really appreciate the information you are giving me.

**Reply by DannyBen** [reply]
> Would you like to be able to run the template and print it to the standard output?

Exactly. I saw the `serve` command, but would also love to be able to do these:

```shell
$ scriggo render template.html   # output to stdout
$ scriggo render template.html output.html  # output to a file
```

of course, the second command is not a hard-must, since it can be done with redirection (`scriggo template.html > out.html`) - but it would still be a nice touch.

**Reply by gazerro** [reply]
I will open two proposals. One to provide compiled binary releases and one to implement a `render` subcommand like the one you suggested.

For binaries we should do it next week. For the `render` subcommand it will take a little longer.

**Reply by DannyBen** [reply]
Excellent, I will keep watch.

**Reply by zapateo** [reply]
The commit https://github.com/open2b/scriggo/commit/b659a905ad2f80a085b29fa18b35d26c5a128567 added support for automatic release of Scriggo binaries using GitHub Actions and [GoReleaser](https://github.com/goreleaser/goreleaser), so the next release of Scriggo will have binaries.

**Reply by DannyBen** [reply]
Sweet!

**Reply by gazerro** [reply]
We have released Scriggo v0.52.0 with scriggo command binaries. 🎉

https://github.com/open2b/scriggo/releases/tag/v0.52.0

**Reply by DannyBen** [reply]
Thank you - I received the GitHub release notification as well. Going to try it later today!
